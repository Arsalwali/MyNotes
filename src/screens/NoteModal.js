import React from 'react';
import {
  StyleSheet,
  View,
  ScrollView,
  Text,
  Modal,
  Button,
  TextInput,
} from 'react-native';
import { connect } from 'react-redux'; 
import { hideModal, setColor } from './actionCreators';
import { addNote, editNote } from '../core/actionCreators/firebase';
import { ColorPicker } from '../components';
import { colors, metrics } from '../theme';

const styles = StyleSheet.create({
  modalContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  modalBox: {
    backgroundColor: colors.white,
    borderRadius: 4,
    padding: 20,
    elevation: 6,
    width: '100%',
  },
  content: {
    flexGrow: 0,
    marginTop: 5,
    marginBottom: 20,
  },
  textInput: {
    height: 40,
    fontSize: metrics.bodyText,
    color: colors.darkText,
    borderColor: colors.gray,
    borderBottomWidth: 1,
  },
  textInputMultiline: {
    color: colors.darkText,
    fontSize: metrics.bodyText,
  },
});

export class NoteModal extends React.Component {
  static getDerivedStateFromProps(props, state) {
    if (props.documentKey !== state.documentKey) {
      return {
        documentKey: props.documentKey,
        title: props.title,
        note: props.note,
        color: props.color,
        isAdd: !props.documentKey ? true : false,
        saveButtonDisable: !props.title ? true: false,
      };
    }
    return null;
  }

  constructor(props) {
    super(props);
    this.state = {
      documentKey: '',
      title: '',
      note: '',
      color: '',
      isAdd: true, 
      titleHasError: false,
      noteHasError: false,
      saveButtonDisable: true,
    };
  }

  handleModalCloseRequest = () => null;

  handlePress = () => {
    if(this.state.isAdd) {
      this.props.addNote({
        key: this.state.documentKey,
        title: this.state.title,
        note: this.state.note,
        color: this.props.color,
      });
    } else {
      this.props.editNote({
        key: this.state.documentKey,
        title: this.state.title,
        note: this.state.note,
        color: this.props.color,
      });
    }
    this.props.hideModal();
  }

  onCancelPress = () => {
    this.setState({ documentKey: '', titleHasError: false, noteHasError: false, saveButtonDisable: true });
    this.props.hideModal();
  }

  getTitleStyling = () => {
    if (!this.state.titleHasError) {
      return styles.textInput;
    }

    return [styles.textInput, { borderColor: colors.error }];
  };

  onTitleChangeTextHandler = (value) => {
    if (value.length === 0) {
      this.setState({
        title: value,
        titleHasError: true,
        saveButtonDisable: true,
      });
      return;
    }

    this.setState({
      title: value,
      titleHasError: false,
      saveButtonDisable: false,
    });
  }

  onNoteChangeText = (value) => {
    if (value.length === 0) {
      this.setState({
        note: value,
        noteHasError: true,
      });
      return;
    }

    this.setState({
      note: value,
      noteHasError: false,
    });
  }

  onColorChange = (color) => {
    if (this.props.setColor) {
      this.props.setColor(color);
    }
  }

  renderContentSection = () => (
    <ScrollView style={styles.content}>
      <ColorPicker color={this.props.color} setColor={this.props.setColor} onColorChange={this.onColorChange} />
      <TextInput
        style={this.getTitleStyling()}
        placeholder="Title"
        value={this.state.title}
        onChangeText={this.onTitleChangeTextHandler}
      />
      <TextInput
        style={styles.textInputMultiline}
        placeholder="Note"
        defaultValue={this.state.note}
        onChangeText={this.onNoteChangeText}
        underlineColorAndroid="transparent"
        multiline
      />
    </ScrollView>
  );

  renderButtonSection = () => (
    <View style={{ flexDirection: 'row', justifyContent: 'space-around'}}>
      <Button title='Cancel' onPress={this.onCancelPress} />
      <Button disabled={this.state.saveButtonDisable} title='Save' onPress={this.handlePress} />
    </View>
  );

  render() {
    return (
      <Modal
        visible={this.props.isVisible}
        animationType="none"
        transparent
        onRequestClose={this.handleModalCloseRequest}
        supportedOrientations={['portrait', 'landscape']}
      >
        <View style={styles.modalContainer}>
          <View style={styles.modalBox}>
            {this.renderContentSection()}
            {this.renderButtonSection()}
          </View>
        </View>
      </Modal>
    );
  }
}

export const mapStateToProps = (state) => {
  const { key, title, note, color, isVisible } = state.modal;
  return { documentKey: key, title, note, color, isVisible };
};

const mapDispatchToProps = { hideModal, setColor, addNote, editNote };

export default connect(mapStateToProps, mapDispatchToProps)(NoteModal);
