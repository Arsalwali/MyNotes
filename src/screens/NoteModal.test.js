import React from 'react';
import { shallow } from 'enzyme';
import { NoteModal, mapStateToProps } from './NoteModal';

describe('Note Modal', () => {
  it('should render properly', () => {
    const props = {
      documentKey: null,
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('getDerivedStateFromProps should return null is document key unchanged', () => {
    const props = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    
    const state = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
    }

    expect(NoteModal.getDerivedStateFromProps(props, state)).toEqual(null);
  });

  it('getDerivedStateFromProps should return new state is document key changed', () => {
    const props = {
      documentKey: '112233',
      title: 'Second Note',
      note: 'Important Note',
      color: '#cff12b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    
    const state = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
    }

    const expectedState = {
      documentKey: '112233',
      title: 'Second Note',
      note: 'Important Note',
      color: '#cff12b',
      isAdd: false,
      saveButtonDisable: false,
    }

    expect(NoteModal.getDerivedStateFromProps(props, state)).toEqual(expectedState);
  });

  it('getDerivedStateFromProps should disable save button if title is empty', () => {
    const props = {
      documentKey: null,
      title: '',
      note: '',
      color: '',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    
    const state = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
    }

    const expectedState = {
      documentKey: null,
      title: '',
      note: '',
      color: '',
      isAdd: true,
      saveButtonDisable: true,
    }

    expect(NoteModal.getDerivedStateFromProps(props, state)).toEqual(expectedState);
  });

  it('handlePress should call addNote if Add is true', () => {
    const props = {
      documentKey: null,
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    expect(props.addNote.mock.calls.length).toBe(0);
    instance.handlePress();
    expect(props.addNote.mock.calls.length).toBe(1);
    expect(props.addNote.mock.calls[0][0]).toEqual({
      key: props.documentKey,
      title: props.title,
      note: props.note,
      color: props.color,
    });
  });

  it('handlePress should call editNote if Add is false', () => {
    const props = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    expect(props.editNote.mock.calls.length).toBe(0);
    instance.handlePress();
    expect(props.editNote.mock.calls.length).toBe(1);
    expect(props.editNote.mock.calls[0][0]).toEqual({
      key: props.documentKey,
      title: props.title,
      note: props.note,
      color: props.color,
    });
  });

  it('handlePress should call hideModal', () => {
    const props = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    expect(props.hideModal.mock.calls.length).toBe(0);
    instance.handlePress();
    expect(props.hideModal.mock.calls.length).toBe(1);
  });

  it('onCancelPress should call hideModal', () => {
    const props = {
      documentKey: '12345',
      title: 'First Note',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    expect(props.hideModal.mock.calls.length).toBe(0);
    instance.onCancelPress();
    expect(props.hideModal.mock.calls.length).toBe(1);
  });

  it('getTitleStyling should retrun error style if title is empty', () => {
    const props = {
      documentKey: '12345',
      title: '',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    instance.setState({ titleHasError: true });
    const errorStyle = instance.getTitleStyling();
    expect(errorStyle).toMatchSnapshot();
  });

  it('getTitleStyling should return simple style if title is filled', () => {
    const props = {
      documentKey: '12345',
      title: 'First',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    const style = instance.getTitleStyling();
    expect(style).toMatchSnapshot();
  });

  it('onTitleChangeTextHandler set errors if title is empty', () => {
    const props = {
      documentKey: '12345',
      title: '',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    instance.onTitleChangeTextHandler('');
    expect(wrapper.state().titleHasError).toEqual(true);
    expect(wrapper.state().saveButtonDisable).toEqual(true);
  });

  it('onTitleChangeTextHandler set errors false if title is filled', () => {
    const props = {
      documentKey: '12345',
      title: '',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    instance.onTitleChangeTextHandler('First');
    expect(wrapper.state().titleHasError).toEqual(false);
    expect(wrapper.state().saveButtonDisable).toEqual(false);
  });

  it('onNoteChangeText set errors if note is empty', () => {
    const props = {
      documentKey: '12345',
      title: 'First',
      note: 'Important Note',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    instance.onNoteChangeText('');
    expect(wrapper.state().noteHasError).toEqual(true);
  });

  it('onNoteChangeText set errors false if title is filled', () => {
    const props = {
      documentKey: '12345',
      title: 'First',
      note: '',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    instance.onNoteChangeText('First');
    expect(wrapper.state().noteHasError).toEqual(false);
  });

  it('onColorChange should call set color if fn is available', () => {
    const props = {
      documentKey: '12345',
      title: 'First',
      note: '',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: jest.fn(),
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    expect(props.setColor.mock.calls.length).toBe(0);
    instance.onColorChange('#c0392b');
    expect(props.setColor.mock.calls.length).toBe(1);
    expect(props.setColor.mock.calls[0][0]).toEqual('#c0392b');
  });

  it('onColorChange should not call set color if fn is not available', () => {
    const props = {
      documentKey: '12345',
      title: 'First',
      note: '',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: null,
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    expect(props.setColor).toBe(null);
    instance.onColorChange('#c0392b');
    expect(props.setColor).toBe(null);
  });

  it('handleModalCloseRequest should return null', () => {
    const props = {
      documentKey: '12345',
      title: 'First',
      note: '',
      color: '#c0392b',
      isVisible: true,
      hideModal: jest.fn(),
      setColor: null,
      addNote: jest.fn(),
      editNote: jest.fn(),
    };
    const wrapper = shallow(<NoteModal {...props} />);
    const instance = wrapper.instance();
    const handle = instance.handleModalCloseRequest();
    expect(handle).toBe(null);
  });

  it('mapStateToProps', () => {
    const state = {
      modal: {
        key: '123',
        title: 'my note',
        note: 'long note',
        color: '#454545',
        isVisible: true,
      },
    };

    const props = mapStateToProps(state);
    expect(props).toEqual({
      documentKey: '123',
      title: 'my note',
      note: 'long note',
      color: '#454545',
      isVisible: true,
    });
  });
});
