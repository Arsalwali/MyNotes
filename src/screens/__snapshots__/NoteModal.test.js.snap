// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Modal getTitleStyling should retrun error style if title is empty 1`] = `
Array [
  Object {
    "borderBottomWidth": 1,
    "borderColor": "#808080",
    "color": "#212121",
    "fontSize": 17.5,
    "height": 40,
  },
  Object {
    "borderColor": "#d50000",
  },
]
`;

exports[`Note Modal getTitleStyling should return simple style if title is filled 1`] = `
Object {
  "borderBottomWidth": 1,
  "borderColor": "#808080",
  "color": "#212121",
  "fontSize": 17.5,
  "height": 40,
}
`;

exports[`Note Modal should render properly 1`] = `
<Component
  animationType="none"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "elevation": 6,
          "padding": 20,
          "width": "100%",
        }
      }
    >
      <ScrollView
        style={
          Object {
            "flexGrow": 0,
            "marginBottom": 20,
            "marginTop": 5,
          }
        }
      >
        <ControlledColorPicker
          color="#c0392b"
          onColorChange={[Function]}
          setColor={[MockFunction]}
        />
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Title"
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#808080",
              "color": "#212121",
              "fontSize": 17.5,
              "height": 40,
            }
          }
          underlineColorAndroid="transparent"
          value="First Note"
        />
        <TextInput
          allowFontScaling={true}
          defaultValue="Important Note"
          multiline={true}
          onChangeText={[Function]}
          placeholder="Note"
          style={
            Object {
              "color": "#212121",
              "fontSize": 17.5,
            }
          }
          underlineColorAndroid="transparent"
        />
      </ScrollView>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <Button
          onPress={[Function]}
          title="Cancel"
        />
        <Button
          disabled={false}
          onPress={[Function]}
          title="Save"
        />
      </View>
    </View>
  </View>
</Component>
`;
